### PR TITLE
feat(drivers): add ONVIF SOAP client with WS-UsernameToken auth

### PR DESCRIPTION
**Pull request by cleithner-comcast**
<sub>Stack created with <a href="https://github.com/github/gh-stack">GitHub Stacks CLI</a> • <a href="https://gh.io/stacks-feedback">Give Feedback 💬</a></sub>

### CODE REVIEW

**Review by rchowdcmcsa — APPROVED** [review]
Just reviewing the code, it looks good apart from a few issues that Copilot AI already flagged.

File: core/deviceDrivers/onvif/OnvifSoapClient.cpp [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+
+namespace barton
+{
+    namespace onvif
+    {
+
+        namespace
+        {
+
+            const char *const NS_SOAP = "http://www.w3.org/2003/05/soap-envelope";
```
Comment (kfundecmcsa) [inline core/deviceDrivers/onvif/OnvifSoapClient.cpp]: what does the `NS` prefix for these variables mean?

File: core/deviceDrivers/onvif/OnvifSoapClient.cpp [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+
+        std::string
+        OnvifComputePasswordDigest(const std::string &nonceRaw, const std::string &created, const std::string &password)
+        {
+            GChecksum *sha1 = g_checksum_new(G_CHECKSUM_SHA1);
+            g_checksum_update(sha1, reinterpret_cast<const guchar *>(nonceRaw.data()), nonceRaw.size());
+            g_checksum_update(sha1, reinterpret_cast<const guchar *>(created.data()), created.size());
+            g_checksum_update(sha1, reinterpret_cast<const guchar *>(password.data()), password.size());
+
+            guint8 digest[20];
```
Comment (kfundecmcsa) [inline core/deviceDrivers/onvif/OnvifSoapClient.cpp]: `20` looks like a magic number, might be good to define this

File: core/deviceDrivers/onvif/OnvifXml.cpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+extern "C" {
+#include <xmlHelper/xmlHelper.h>
+}
+
+namespace barton
+{
+    namespace onvif
+    {
+
+        std::string OnvifXmlEscape(const std::string &in)
```
Comment (kfundecmcsa) [inline core/deviceDrivers/onvif/OnvifXml.cpp]: we already pull in libxml, so you may want to consider using `libxml2`'s `xmlEncodeSpecialChars` rather than crafting our own escape function

File: core/test/CMakeLists.txt [hunk 4]
```diff
@@ -420,3 +420,29 @@ if (BCORE_OBSERVABILITY_BACKEND STREQUAL "memory")
...
+        SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/src/OnvifSoapClientTest.cpp
+                ${ONVIF_DIR}/OnvifSoapClient.cpp
+                ${ONVIF_DIR}/OnvifXml.cpp
+        LIBS ${ONVIF_TEST_CURL_LINK_LIBRARIES} ${ONVIF_TEST_XML_LINK_LIBRARIES} gmock BartonCommon::xhXmlHelper
+        INCLUDES ${ONVIF_DIR} ${ONVIF_TEST_XML_INCLUDE_DIRS} ${ONVIF_TEST_CURL_INCLUDE_DIRS}
+    )
+
+    if (BUILD_TESTING)
+        bcore_configure_glib()
+    endif()
```
[the remark this answers is not in the document]
Comment (kfundecmcsa) [reply core/test/CMakeLists.txt]: agreed with copilot

**Review by tleacmcsa — COMMENTED** [review]
I reviewed this PR, which introduces a self-contained ONVIF SOAP client (libcurl + libxml2 + glib) with unit tests. This has already been through extensive review rounds (visible in the commit history) that addressed real security issues:

- one-time `curl_global_init` via `std::call_once`
- `tokensOut.clear()` to avoid stale accumulation
- stripping embedded userinfo from stream/snapshot URIs (credential leak)
- `XML_PARSE_NONET` on all parsers (XXE/network-fetch hardening)
- null-safe XML text/attr helpers
- `responseOut.clear()` before each POST (no response concatenation)
- CSPRNG (`getrandom`) instead of glib's non-cryptographic PRNG for the WS-Security nonce
- only emitting the WS-Security header when both username and password are set
- rejecting SOAP Fault/empty responses as failed `GetDeviceInformation` parses

Given all that, the code is in good shape. A few remaining observations for this round:

**Correctness / robustness**
- `OnvifSoapClient.cpp:279` — `CURLOPT_TIMEOUT` bounds the whole transfer, so a slow/unreachable host is bounded, but there's no `CURLOPT_CONNECTTIMEOUT` set explicitly; not a blocker since `CURLOPT_TIMEOUT` covers connect+transfer, just worth confirming that's the intended behavior for camera discovery timeouts.
- No `CURLOPT_SSL_VERIFYPEER/VERIFYHOST` handling for `https://` service URLs. Most ONVIF cameras are plain HTTP, but if this client is ever pointed at an `https` endpoint with a self-signed cert (common for cheap cameras), it will fail closed with default libcurl verification. Worth a comment/TODO if HTTPS cameras are in scope, since there's currently no way to configure this from `OnvifSoapClient`.
- `StripUriCredentials` (OnvifSoapClient.cpp:163) does a naive `find('@')`/`find('/')`; it won't handle IPv6 literal hosts (`rtsp://user:pass@[::1]:554/...`) correctly (the `@` is still before the first `/`, so it should still work, but bracketed-host edge cases weren't tested). Low risk given typical camera URIs, but could use a test case.

**Minor/style**
- `core/test/CMakeLists.txt:766` and `:785` — the outer `if (BCORE_ONVIF AND BUILD_TESTING)` already guards the whole block, so the inner `if (BUILD_TESTING) bcore_configure_glib() endif()` is redundant (always true inside the outer block). Harmless, but could be simplified to a plain call.
- `OnvifSoapClient::GetStreamUri`/`GetSnapshotUri` share almost identical bodies (build request, escape token, post, parse media URI, error message) — could be factored into a private helper taking the operation name/namespace, but this is a nit, not a defect.

**Tests**
- Good coverage: password digest (known vector + determinism), envelope with/without credentials (and asserts the cleartext password never appears — nice), device-info parsing (including the SOAP-Fault-rejection case), profile token parsing (including the empty-profiles failure case), and stream/snapshot URI parsing including the credential-stripping case.
- No test exercises `OnvifSoapClient::Post`/the full `GetDeviceInformation` etc. over real HTTP (reasonable to skip — would need a mock server — but worth noting there's no coverage of the HTTP-status ≥400 error path or transport-error path at all, only the pure parsing/envelope-building functions are tested).

Overall: solid, already hardened through prior review rounds. Nothing blocking; the HTTPS/cert-verification gap and the missing HTTP-status/transport-error test coverage are the two things I'd flag before merge, though neither is critical for an HTTP-only ONVIF deployment.
